### PR TITLE
feat: Added bindings

### PR DESCRIPTION
**Pull request by robertleeplummerjr**
First go on
* glBufferSubData
* glUniform2f
* glUniform2iv
* glUniform3iv

---
This change is [Reviewable](https://reviewable.io/reviews/google/node-gles/4)

### CODE REVIEW

**Comment by robertleeplummerjr** [comment]
I signed it!

**Comment by robertleeplummerjr** [comment]
no no, thank you @googlebot.

File: binding/webgl_rendering_context.cc [hunk 1]
```diff
@@ -224,6 +224,7 @@ napi_status WebGLRenderingContext::Register(napi_env env, napi_value exports) {
 // blendFunc(sfactor: number, dfactor: number): void;
 // blendFuncSeparate(srcRGB: number, dstRGB: number, srcAlpha: number, dstAlpha: number): void;
       NAPI_DEFINE_METHOD("bufferData", BufferData),
+      NAPI_DEFINE_METHOD("bufferSubData", BufferSubData),
 // bufferSubData(target: number, offset: number, data: Int8Array | Int16Array | Int32Array | Uint8Array | Uint16Array | Uint32Array | Uint8ClampedArray | Float32Array | Float64Array | DataView | ArrayBuffer | null): void;
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: Delete this line here.

File: binding/webgl_rendering_context.cc [hunk 2]
```diff
@@ -347,15 +348,15 @@ napi_status WebGLRenderingContext::Register(napi_env env, napi_value exports) {
       NAPI_DEFINE_METHOD("uniform1f", Uniform1f),
       NAPI_DEFINE_METHOD("uniform1fv", Uniform1fv),
       NAPI_DEFINE_METHOD("uniform1i", Uniform1i),
+      NAPI_DEFINE_METHOD("uniform2f", Uniform2f),
 // uniform1iv(location: WebGLUniformLocation | null, v: Int32Array | ArrayLike<number>): void;
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: Ditto with these - this is my rough  TODO list. Once I implement a method I delete the commented out declaration. Please do that for all others.

File: binding/webgl_rendering_context.cc [hunk 3]
```diff
@@ -986,6 +987,73 @@ napi_value WebGLRenderingContext::BufferData(napi_env env,
...
+  WebGLRenderingContext *context = nullptr;
+  nstatus = UnwrapContext(env, js_this, &context);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  ENSURE_VALUE_IS_NUMBER_RETVAL(env, args[0], nullptr);
+  GLenum target;
+  nstatus = napi_get_value_uint32(env, args[0], &target);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  // start here
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: Delete this comment.

File: binding/webgl_rendering_context.cc [hunk 4]
```diff
@@ -986,6 +987,73 @@ napi_value WebGLRenderingContext::BufferData(napi_env env,
...
+  napi_value js_this;
+  nstatus = napi_get_cb_info(env, info, &argc, args, &js_this, nullptr);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+  ENSURE_ARGC_RETVAL(env, argc, 3, nullptr);
+
+  WebGLRenderingContext *context = nullptr;
+  nstatus = UnwrapContext(env, js_this, &context);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  ENSURE_VALUE_IS_NUMBER_RETVAL(env, args[0], nullptr);
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: Nit - I generally try and do all the js-arg validation at once:

```c++
ENSURE_VALUE_IS_NUMBER_RETVAL(env, args[0], nullptr);
ENSURE_VALUE_IS_NUMBER_RETVAL(env, args[1], nullptr);
ENSURE_VALUE_IS_NUMBER_RETVAL(env, args[2], nullptr);
...
```

File: binding/webgl_rendering_context.cc [hunk 5]
```diff
@@ -2466,6 +2576,82 @@ napi_value WebGLRenderingContext::Uniform2i(napi_env env,
...
+
+  GLint location;
+  nstatus = napi_get_value_int32(env, args[0], &location);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  size_t size;
+  void *data;
+  nstatus = napi_get_typedarray_info(env, args[1], nullptr, &size, &data,
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: You'll need to check the type here. According to the specs - value can also be a regular array:

https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/uniform

File: binding/webgl_rendering_context.cc [hunk 6]
```diff
@@ -2466,6 +2576,82 @@ napi_value WebGLRenderingContext::Uniform2i(napi_env env,
...
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: Ditto here about type checking. A user can pass in a regular array as well as a typed array for values.

File: binding/webgl_rendering_context.cc [hunk 7]
```diff
@@ -986,6 +987,73 @@ napi_value WebGLRenderingContext::BufferData(napi_env env,
...
+  uint32_t offset;
+  ENSURE_VALUE_IS_NUMBER_RETVAL(env, args[1], nullptr);
+  nstatus = napi_get_value_uint32(env, args[1], &offset);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  napi_valuetype arg_valuetype;
+  nstatus = napi_typeof(env, args[2], &arg_valuetype);
+  ENSURE_NAPI_OK_RETVAL(env, nstatus, nullptr);
+
+  if (arg_valuetype == napi_object) {
```
Comment (nkreeger) [inline binding/webgl_rendering_context.cc]: gl.bufferSubData() only takes ArrayView-instances:

https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/bufferSubData

You can check with `napi_is_arraybufferview` and `napi_get_arraybuffer_info`. Feel free to add a macro to utils.h for `ENSURE_NAPI_IS_VALUE_ARRAY_BUFFER*`

File: binding/webgl_rendering_context.cc (hunk 5 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: My gut was to include this originally, but I didn't see (naive eye here) any usages.  Can you point me to one so I can emulate?

Comment (nkreeger) [reply binding/webgl_rendering_context.cc]: You'll need to check for "is this an object" and "is this a typed array?". You'll also need to check for if this is an array.

Pseudo-code:
```c++

bool is_array = false;
napi_is_array(..., args[1], &is_arary, ...);
if (is_array) {
  // Extract buffer:
} else {
  //  Check for typed array:
  
}

```

Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: >`nstatus = napi_get_arraybuffer_info(env, args[1], &data, &size);`

When I try this using a standard array, it fails, as I believe `napi_get_arraybuffer_info` is trying to handle a buffer, but what we have is actually a plain javascript array.

**Comment by nkreeger** [comment]
Sorry for the lag - work was busy with other projects and I have some time
to get this working this week...

Let's add a `TODO` for supporting regular JS arrays.

**Comment by nkreeger** [comment]
@robertleeplummerjr I merged my large working branch today. If you don't mind syncing this when you get a chance.

File: binding/webgl_rendering_context.cc (hunk 3 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: removed

File: binding/webgl_rendering_context.cc (hunk 4 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: Reformatted.  Ty for the pointer!

File: binding/webgl_rendering_context.cc (hunk 5 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: Added `TODO` to satisfy comment from https://github.com/google/node-gles/pull/4#issuecomment-458769624

File: binding/webgl_rendering_context.cc (hunk 6 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: Added `TODO`

File: binding/webgl_rendering_context.cc (hunk 7 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: Added `ENSURE_VALUE_IS_ARRAY_BUFFER` and `ENSURE_VALUE_IS_ARRAY_BUFFER_RETVAL` to utils.h.

File: binding/webgl_rendering_context.cc (hunk 2 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: I can delete, but "uniform2f" !== "uniform1iv"

File: binding/webgl_rendering_context.cc (hunk 1 above)
Comment (robertleeplummerjr) [reply binding/webgl_rendering_context.cc]: Deleted.  Ty!

File: binding/webgl_rendering_context.cc (hunk 6 above)
Comment (nkreeger) [reply binding/webgl_rendering_context.cc]: Can you add a check for object/typedarray like you did in `::BufferSubData()`?

File: binding/webgl_rendering_context.cc (hunk 5 above)
Comment (nkreeger) [reply binding/webgl_rendering_context.cc]: Ditto here about type checking -

**Comment by nkreeger** [comment]
Thanks @robertleeplummerjr I'm merging this into one of my working branches and I'll drive changes from here!

**Comment by robertleeplummerjr** [comment]
ty sir

**Comment by nkreeger** [comment]
Just published node-gles 0.0.5 with your changes.

**Pull request merged**